File: SecondYear/semestrul4/MPP/examen_practic/Rezolvari/generic/networking/src/main/java/org/networking/utils/AbstractServer.java
package org.networking.utils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.rmi.ServerException;

public abstract class AbstractServer {
    private int port;
    private ServerSocket server = null;
    private static final Logger logger = LogManager.getLogger(AbstractServer.class);


    public AbstractServer(int port) {
        this.port = port;
    }

    public void start() throws ServerException {
        try {
            server = new ServerSocket(port);
            logger.info("Server started on port {}", port);
            while (true) {
                logger.info("Waiting for clients ...");
                Socket client = server.accept();
                logger.info("Client connected: {}", client.getInetAddress());
                processRequest(client);
            }
        } catch (IOException e) {
            logger.error("Error starting server", e);
            throw new ServerException("Starting server error ", e);
        }
    }

    protected abstract void processRequest(Socket client);

    public void stop() throws ServerException {
        try {
            if (server != null && !server.isClosed()) {
                server.close();
                logger.info("Server stopped.");
            }
        } catch (IOException e) {
            logger.error("Error stopping server", e);
            throw new ServerException("Closing server error ", e);
        }
    }
}

File: SecondYear/semestrul4/MPP/PROIECT/CSharpProjectNetworking/Networking/utils/AbstractServer.cs

using log4net;
using System.Net;
using System.Net.Sockets;

namespace Networking.utils
{
    public abstract class AbstractServer
    {
        private int port;
        private TcpListener server;
        private static readonly ILog log = LogManager.GetLogger(typeof(AbsConcurrentServer));

        public AbstractServer(int port)
        {
            this.port = port;
        }

        public void Start()
        {
            try
            {
                server = new TcpListener(IPAddress.Any, port);
                server.Start();
                log.Info("Server started on port {port}" + port);
                while (true)
                {
                    log.Info("Waiting for clients...");
                    TcpClient client = server.AcceptTcpClient();
                    log.Info("Client connected...");
                    ProcessRequest(client);
                }
            }
            catch (Exception e)
            {
                log.Error(e);
                throw new ServerException("Starting server error", e);
            }
        }

        public void Stop()
        {
            try
            {
                server?.Stop();
                log.Info("Server stopped.");
            }
            catch (Exception e)
            {
                log.Error(e + "Error stopping server");
                throw new ServerException("Closing server error", e);
            }
        }

        protected abstract void ProcessRequest(TcpClient client);
    }

}
